File: library/src/test/java/io/ashdavies/rx/rxtasks/CompletableTaskOnSubscribeTest.java
package io.ashdavies.rx.rxtasks;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import io.reactivex.CompletableEmitter;
import io.reactivex.CompletableOnSubscribe;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

@RunWith(MockitoJUnitRunner.class)
public class CompletableTaskOnSubscribeTest {

  private CompletableOnSubscribe onSubscribe;

  @Mock Task<Void> task;
  @Mock CompletableEmitter emitter;

  @Mock TaskListenerFactory<Void, CompletableEmitter> factory;
  @Mock OnSuccessListener<Void> onSuccessListener;
  @Mock OnFailureListener onFailureListener;

  @Before
  public void setUp() throws Exception {
    onSubscribe = new CompletableTaskOnSubscribe(task, factory);
  }

  @Test
  public void shouldSubscribeWithOnSuccessListener() throws Exception {
    given(factory.createOnSuccessListener(emitter)).willReturn(onSuccessListener);

    onSubscribe.subscribe(emitter);

    then(factory).should().createOnSuccessListener(emitter);
    then(task).should().addOnSuccessListener(onSuccessListener);
  }

  @Test
  public void shouldSubscribeWithOnFailureListener() throws Exception {
    given(factory.createOnFailureListener(emitter)).willReturn(onFailureListener);

    onSubscribe.subscribe(emitter);

    then(factory).should().createOnFailureListener(emitter);
    then(task).should().addOnFailureListener(onFailureListener);
  }
}

File: library/src/test/kotlin/io/ashdavies/rx/rxtasks/CompletableTaskOnSubscribeTest.kt
package io.ashdavies.rx.rxtasks

import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.Task
import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.given
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.then
import io.reactivex.CompletableEmitter
import org.junit.Test
import java.util.concurrent.Executor

internal class CompletableTaskOnSubscribeTest {

  private val task = mock<Task<Void>>()
  private val emitter = mock<CompletableEmitter>()
  private val executor = mock<Executor>()

  private val completable = CompletableTaskOnSubscribe(task, executor)

  @Test
  fun `should add on success listener`() {
    val captor = argumentCaptor<OnSuccessListener<Void>>()
    given(task.addOnSuccessListener(eq(executor), captor.capture())).willReturn(task)

    completable.subscribe(emitter)
    captor.lastValue.onSuccess(null)

    then(emitter).should().onComplete()
  }

  @Test
  fun `should add on failure listener`() {
    val captor = argumentCaptor<OnFailureListener>()
    given(task.addOnFailureListener(eq(executor), captor.capture())).willReturn(task)

    val exception = RuntimeException()
    completable.subscribe(emitter)

    captor.lastValue.onFailure(exception)
    then(emitter).should().onError(exception)
  }
}
